File: june18exercise.js
// # Given an array of strings and an integer k, your task is to return the longest string consisting of k consecutive strings from the array concatenated together.
// # longestConsecutive(['hi', 'marbles', 'mittens', 'bye', 'lorem', 'ipsum', 'to', 'a', 'hippocampus'], 3); // -> 'marblesmittensbye'
// # If there is a tie, return the first one.
// # longestConsecutive(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], 2); // --> "abigailtheta"
// # Return "" if the array is empty or if k is negative or larger than the length of the array.

function longestConsecutive(array, integer) {
    if (array.length === 0 || integer > array.length || integer < 0) {
        console.log('""');
    } else {
        const sortedArray = array.sort((a, b) => b.length - a.length)
        const sortedSet = new Set(sortedArray)
        const setArray = Array.from(sortedSet)
        longest = []
        for (i=0;i<integer;i++) {
            longest.push(setArray[i])
        }
        final = longest.join('')
        console.log(final)
    }
}


longestConsecutive(['hi', 'marbles', 'mittens', 'bye', 'lorem', 'ipsum', 'to', 'a', 'hippocampus'], 3)
longestConsecutive(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], 2); 
longestConsecutive([], 2); 
longestConsecutive(['hi'], 2); 
longestConsecutive(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], -2); 


// The above solution works to get the output as given in the assignment
// BUT for longestConsecutive(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], 2);
// We might actually expect that the longest string (considering the duplicates) would be abagailabagail
// And not abagailtheta
// The solution for considering duplicates is posted below:

function longestConsecutiveExtended(array, integer) {
    if (array.length === 0 || integer > array.length || integer < 0) {
        console.log('""');
    } else {
        const sortedArray = array.sort((a, b) => b.length - a.length)
        longest = []
        for (i=0;i<integer;i++) {
            longest.push(sortedArray[i])
        }
        final = longest.join('')
        console.log(final)
    }
}

longestConsecutiveExtended(['hi', 'marbles', 'mittens', 'bye', 'lorem', 'ipsum', 'to', 'a', 'hippocampus'], 3)
longestConsecutiveExtended(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], 2); 
longestConsecutiveExtended([], 2); 
longestConsecutiveExtended(['hi'], 2); 
longestConsecutiveExtended(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], -2); 

File: june18exercise.py
# Given an array of strings and an integer k, your task is to return the longest string consisting of k consecutive strings from the array concatenated together.
# longestConsecutive(['hi', 'marbles', 'mittens', 'bye', 'lorem', 'ipsum', 'to', 'a', 'hippocampus'], 3); // -> 'marblesmittensbye'
# If there is a tie, return the first one.
# longestConsecutive(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], 2); // --> "abigailtheta"
# Return "" if the array is empty or if k is negative or larger than the length of the array.


def longestConsecutive(array, integer):
    if array == [] or integer > len(array) or integer < 0:
        print('""')
    else:
        words_and_lengths = []
        # if you do not want duplicates concatenated (only unique words)
        for word in set(array):
        # if you are okay with duplicates being concatenated
        # for word in array:
            words_and_lengths.append({'word': word, 'length': len(word)})
        ordered_words_and_lengths = sorted(words_and_lengths, key=lambda k: k['length'], reverse=True)
        longest_words = ordered_words_and_lengths[0:integer]
        top_words = []
        for item in longest_words:
            top_words.append(item['word'])
        print(''.join(top_words))


longestConsecutive(['hi', 'marbles', 'mittens', 'bye', 'lorem', 'ipsum', 'to', 'a', 'hippocampus'], 3)

# with this example, the longest consecutive string would be abagailabagail unless we only want unique values
# this is why there is an option for either line 14 or line 16 above
longestConsecutive(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], 2); 

longestConsecutive([], 2); 

longestConsecutive(['hi'], 2); 

longestConsecutive(["zone", "abigail", "theta", "form", "libe", "zas", "theta", "abigail"], -2); 
